### DISCUSSION TITLE
Changing the server address more than once

### DISCUSSION THREAD

**Discussion by dboycepa** [activexray/tftp_client#2] [General]
@kiranshila, thanks for the prompt action, the changes for 0.2.0 to fix Issue #1.  Before you issue the new release, I have some observations which will improve the fix.

IMO, the upload() assignment at https://github.com/kiranshila/tftp_client/blob/3101eb96daefa1a77db06a0fc62def0ffb0b16eb/src/lib.rs#L199 is in the wrong place, and introduces a vulnerability for highjacking the transfer.  The problem is three-fold: 

1. an incoming udp message to the client port from _any_ remote machine can change the whole server address (IP and port), regardless of whether it's even a TFTP message;
2. the address of the remote server should not change at all, only the port;
3. the protocol spec you quoted indicates the remote port should only change once.
> `The two chosen TID's are then used for the remainder of the transfer.`

Instead, consider making a port-only-update at https://github.com/kiranshila/tftp_client/blob/3101eb96daefa1a77db06a0fc62def0ffb0b16eb/src/lib.rs#L235 , where a) you know it's a TFTP ACK packet and b) it's the first response.

I'd also recommend checking that at https://github.com/kiranshila/tftp_client/blob/3101eb96daefa1a77db06a0fc62def0ffb0b16eb/src/lib.rs#L226 you silently drop any packet for which the remote socket address does not match the server packet address, and also drop any packet for which the remote port does not match (making an exception for what may be the first ACK packet, which will have a different port).

Similar considerations apply to the download() code.

Hopefully this will help make the tftp_client more robust.

**Comment by activexray** [comment]
Great suggestions! I'll implement these for the next release

**Comment by activexray** [comment]
Sorry this took so long, I've been finishing my PhD :)

Can you check out the code in the async_native branch?
